### PR TITLE
Small improvements to nicer CI errors

### PR DESCRIPTION
**Pull request by BD103**
This refactors a few bits of code, but does not change any logic. :)

I recommend viewing the changes one commit at a time, it's easier to digest what's happening!

### CODE REVIEW

File: tools/ci/src/commands/mod.rs [hunk 1]
```diff
@@ -115,15 +124,11 @@ pub fn run_cargo_command_with_json(
...
-            Err(())
-        }
-    }
+    // Don't just leave the child running in the background.
+    let _ = child.wait();
+
+    JsonCommandOutput::from_cargo_output(command_out, command_name.to_string()).map_err(|err| {
+        // Log error, converting it to an `Err(())`.
+        eprintln!("{err}");
+    })
```
Comment (Brezak) [inline tools/ci/src/commands/mod.rs]: Did you tests this? I'm pretty sure this deadlocks.
1. Cargo writes a bunch of data to it's stdout.
2. The pipe buffer gets filled. All subsequent writes block.
3. We wait for the process to finish.
4. Process is stuck waiting for stdout write to finish.

The old behavior would be:
1. Cargo writes a bunch of data to it's stdout.
2. We consume this data in `from_cargo_output`.
2.a If parsing the data fails we drop the child stdout handle closing it. All subsequent writes will fail.
3. We wait for the child to finish.

File: tools/ci/src/commands/mod.rs [hunk 2]
```diff
@@ -96,16 +101,20 @@ pub fn run_cargo_command_with_json(
...
+            "cargo",
+            cargo_command,
+            // Here we specify that we want JSON output.
+            "--message-format",
+            "json",
+        ])
         .args(flags)
         .envs(env.iter().copied())
+        // TODO: What's this line doing?
         .stdout(Stdio::piped())
```
Comment (Brezak) [inline tools/ci/src/commands/mod.rs]: We need to read what cargo outputs while it's running.
The childs stdio is configured as such:
- `stdin` is inherited from ours. Any reads cargo attempts will be read from our `stdin`.
- `stderr` is inherited from ours. Any writes will show up in console. Due to how cargo determines if output should be colored we should also get color on this output.
- `stdout` is piped directly to us. Any writes to it go directly to a [handle](https://doc.rust-lang.org/std/process/struct.ChildStdout.html) we can [read](https://doc.rust-lang.org/std/io/trait.Read.html). We handle this data in `JsonCommandOutput::from_cargo_output`

**Review by Brezak — CHANGES_REQUESTED** [review]

File: tools/ci/src/commands/mod.rs (hunk 2 above)
Comment (BD103) [reply tools/ci/src/commands/mod.rs]: Thank you, that makes a lot more sense! (I misread `Stdio` as `Stdin`, and was wondering why we were connecting `stdout` and `stdin` together 😅)

File: tools/ci/src/commands/mod.rs (hunk 1 above)
Comment (BD103) [reply tools/ci/src/commands/mod.rs]: > Did you tests this? I'm pretty sure this deadlocks.

I didn't test this (probably should have!), and it definitely deadlocks. Let me revert that!

**Pull request merged**